diff --git a/projects/hipblaslt/library/src/amd_detail/rocblaslt/src/include/rocroller_host_internal.hpp b/projects/hipblaslt/library/src/amd_detail/rocblaslt/src/include/rocroller_host_internal.hpp
index e37001eccf..62d4d9f354 100644
--- a/projects/hipblaslt/library/src/amd_detail/rocblaslt/src/include/rocroller_host_internal.hpp
+++ b/projects/hipblaslt/library/src/amd_detail/rocblaslt/src/include/rocroller_host_internal.hpp
@@ -180,11 +180,17 @@ constexpr auto generateTileList() {
 
         int unroll = preferredUnrolling(typeA, typeB, wgt);
 
+        int default_wgm = 6;
+        int non_temporal_a = 0;
+        int non_temporal_b = 0;
+
         tileList[i] = std::make_tuple(
             wgt.m, wgt.n, wgtk * unroll,
             MI.m, MI.n, MI.k,
             1, // occupancy
-            6 // WGM
+            default_wgm, // WGM
+            non_temporal_a, // nontemporal_a
+            non_temporal_b // nontemporal_b
         );
     }
 
diff --git a/projects/hipblaslt/tensilelite/Tensile/Contractions.py b/projects/hipblaslt/tensilelite/Tensile/Contractions.py
index 8291d4ac8e..0c8d8aa6cd 100644
--- a/projects/hipblaslt/tensilelite/Tensile/Contractions.py
+++ b/projects/hipblaslt/tensilelite/Tensile/Contractions.py
@@ -592,7 +592,9 @@ class SizeMapping:
                  'CUOccupancy',
                  'PrefetchGlobalRead',
                  'MathClocksUnrolledLoop',
-                 'synchronizerSizePerWG'
+                 'synchronizerSizePerWG',
+                 'nonTemporalA',
+                 'nonTemporalB',
                  ]
 
     @classmethod
@@ -644,7 +646,9 @@ class SizeMapping:
                    CUOccupancy              = d['CUOccupancy'],
                    PrefetchGlobalRead       = pgr,
                    MathClocksUnrolledLoop   = d['MathClocksUnrolledLoop'],
-                   synchronizerSizePerWG    = synchronizerSizePerWG
+                   synchronizerSizePerWG    = synchronizerSizePerWG,
+                   nonTemporalA             = d['NonTemporalA'],
+                   nonTemporalB             = d['NonTemporalB'],
                    )
     @classmethod
     def ReadOriginalMacroTile(cls, d):
diff --git a/projects/hipblaslt/tensilelite/include/Tensile/ContractionSolution.hpp b/projects/hipblaslt/tensilelite/include/Tensile/ContractionSolution.hpp
index c3815c3f9f..4445e378af 100644
--- a/projects/hipblaslt/tensilelite/include/Tensile/ContractionSolution.hpp
+++ b/projects/hipblaslt/tensilelite/include/Tensile/ContractionSolution.hpp
@@ -158,6 +158,9 @@ namespace TensileLite
         int MathClocksUnrolledLoop = 0;
 
         size_t synchronizerSizePerWG = 0;
+
+        int nonTemporalA = 0;
+        int nonTemporalB = 0;
     };
 
     struct StreamKSettings
diff --git a/projects/hipblaslt/tensilelite/include/Tensile/PredictionLibrary.hpp b/projects/hipblaslt/tensilelite/include/Tensile/PredictionLibrary.hpp
index 5f499e3b9a..c7271f1568 100644
--- a/projects/hipblaslt/tensilelite/include/Tensile/PredictionLibrary.hpp
+++ b/projects/hipblaslt/tensilelite/include/Tensile/PredictionLibrary.hpp
@@ -200,7 +200,9 @@ namespace TensileLite
                                                               std::get<5>(tile),
                                                               std::get<6>(tile),
                                                               std::get<7>(tile),
-                                                              std::get<8>(tile)
+                                                              std::get<8>(tile),
+                                                              std::get<9>(tile),
+                                                              std::get<10>(tile)
                                                             ));
                 auto smapiter = solutionmap.find(mapiter->second);
                 if(mapiter != tile_map.end() && smapiter != solutionmap.end())
diff --git a/projects/hipblaslt/tensilelite/include/Tensile/Serialization/ContractionSolution.hpp b/projects/hipblaslt/tensilelite/include/Tensile/Serialization/ContractionSolution.hpp
index 4560a72f0a..719a74aa97 100644
--- a/projects/hipblaslt/tensilelite/include/Tensile/Serialization/ContractionSolution.hpp
+++ b/projects/hipblaslt/tensilelite/include/Tensile/Serialization/ContractionSolution.hpp
@@ -126,6 +126,8 @@ namespace TensileLite
                 iot::mapRequired(io, "PrefetchGlobalRead", s.PrefetchGlobalRead);
                 iot::mapRequired(io, "MathClocksUnrolledLoop", s.MathClocksUnrolledLoop);
                 iot::mapRequired(io, "synchronizerSizePerWG", s.synchronizerSizePerWG);
+                iot::mapRequired(io, "nonTemporalA", s.nonTemporalA);
+                iot::mapRequired(io, "nonTemporalB", s.nonTemporalB);
             }
 
             const static bool flow = false;
diff --git a/projects/hipblaslt/tensilelite/include/Tensile/Serialization/PredictionLibrary.hpp b/projects/hipblaslt/tensilelite/include/Tensile/Serialization/PredictionLibrary.hpp
index 717d30c3a1..77da8d810c 100644
--- a/projects/hipblaslt/tensilelite/include/Tensile/Serialization/PredictionLibrary.hpp
+++ b/projects/hipblaslt/tensilelite/include/Tensile/Serialization/PredictionLibrary.hpp
@@ -84,14 +84,17 @@ namespace TensileLite
                             lib.solutionmap.insert(std::make_pair(index, solution));
 
                             auto solution_tuple = std::make_tuple(
-                                solution->sizeMapping.macroTile.x, // MT_M
-                                solution->sizeMapping.macroTile.y, // MT_N
-                                solution->sizeMapping.depthU, // MT_K
+                                solution->sizeMapping.macroTile.x,          // MT_M
+                                solution->sizeMapping.macroTile.y,          // MT_N
+                                solution->sizeMapping.depthU,               // MT_K
                                 solution->sizeMapping.matrixInstruction[0], // MI_M
                                 solution->sizeMapping.matrixInstruction[1], // MI_N
                                 solution->sizeMapping.matrixInstruction[2], // MI_K
-                                solution->sizeMapping.CUOccupancy,      // Occupancy
-                                solution->sizeMapping.workGroupMapping); // WGM
+                                solution->sizeMapping.CUOccupancy,          // Occupancy
+                                solution->sizeMapping.workGroupMapping,     // WGM
+                                solution->sizeMapping.nonTemporalA,         // Cache flag: A
+                                solution->sizeMapping.nonTemporalB          // Cache flag: B
+                            ); 
 
                             lib.tile_list.emplace_back(solution_tuple);
                             lib.tile_map.insert(std::make_pair(solution_tuple, index));
diff --git a/shared/origami/include/origami/utils.hpp b/shared/origami/include/origami/utils.hpp
index 6dce2c6506..6101462372 100644
--- a/shared/origami/include/origami/utils.hpp
+++ b/shared/origami/include/origami/utils.hpp
@@ -20,7 +20,9 @@ namespace origami
                                        size_t, // MI_N
                                        size_t, // MI_K
                                        size_t,  // Occupancy
-                                       int>;    // WGM
+                                       int,     // WGM
+                                       size_t, // non_temporal_a
+                                       size_t>; // non_temporal_b
 
         using tile_tuple = std::tuple<size_t, // MT_M
                                      size_t, // MT_N
@@ -29,7 +31,9 @@ namespace origami
                                      size_t, // MI_N
                                      size_t, // MI_K
                                      size_t,  // Occupancy
-                                     int>;    // WGM
+                                     int,     // WGM
+                                     size_t, // non_temporal_a
+                                     size_t>; // non_temporal_b
 
         size_t select_best_grid_size(size_t          M,
                                      size_t          N,
diff --git a/shared/origami/src/origami/utils.cpp b/shared/origami/src/origami/utils.cpp
index a58f549934..04882df03d 100644
--- a/shared/origami/src/origami/utils.cpp
+++ b/shared/origami/src/origami/utils.cpp
@@ -213,14 +213,16 @@ namespace origami
 
             for(const auto& mt : MT_list)
             {
-                size_t MT_M      = std::get<0>(mt);
-                size_t MT_N      = std::get<1>(mt);
-                size_t MT_K      = std::get<2>(mt);
-                size_t MI_M      = std::get<3>(mt);
-                size_t MI_N      = std::get<4>(mt);
-                size_t MI_K      = std::get<5>(mt);
-                size_t occupancy = std::get<6>(mt);
-                size_t WGM       = std::get<7>(mt);
+                size_t MT_M             = std::get<0>(mt);
+                size_t MT_N             = std::get<1>(mt);
+                size_t MT_K             = std::get<2>(mt);
+                size_t MI_M             = std::get<3>(mt);
+                size_t MI_N             = std::get<4>(mt);
+                size_t MI_K             = std::get<5>(mt);
+                size_t occupancy        = std::get<6>(mt);
+                size_t WGM              = std::get<7>(mt);
+                size_t non_temporal_a   = std::get<8>(mt);
+                size_t non_temporal_b   = std::get<9>(mt);
 
                 if(debug)
                 {
@@ -254,7 +256,7 @@ namespace origami
                                                                  debug);
 
                     valid_results.emplace_back(
-                        Total_latency, MT_M, MT_N, MT_K, MI_M, MI_N, MI_K, occupancy, WGM);
+                        Total_latency, MT_M, MT_N, MT_K, MI_M, MI_N, MI_K, occupancy, WGM, non_temporal_a, non_temporal_b);
                 }
                 else if(debug)
                 {
@@ -311,9 +313,11 @@ namespace origami
                 {
                     std::cout << M << "x" << N << "x" << K
                               << "Selected Macro-Tile: Latency=" << std::get<0>(tile)
-                              << ", MT_M=" << std::get<1>(tile) << ", MT_N=" << std::get<2>(tile)
-                              << ", MT_K=" << std::get<3>(tile) << ", MI_M=" << std::get<4>(tile)
-                              << ", MI_N=" << std::get<5>(tile) << ", MI_K=" << std::get<6>(tile)
+                              << ", MT_M=" << std::get<0>(tile) << ", MT_N=" << std::get<1>(tile)
+                              << ", MT_K=" << std::get<2>(tile) << ", MI_M=" << std::get<3>(tile)
+                              << ", MI_N=" << std::get<4>(tile) << ", MI_K=" << std::get<5>(tile)
+                              << ", Occupancy=" << std::get<6>(tile) << ", WGM=" << std::get<7>(tile)
+                              << ", NonTemporalA=" << std::get<8>(tile) << ", NonTemporalB=" << std::get<9>(tile)
                               << "\n";
                 }
             }
